fedimint-core/src/task/jit.rs

```diff
@@ -0,0 +1,208 @@
+    _cancel_tx: sync::mpsc::Sender<()>,
+    // since joinhandles, are not portable, we use this to wait for the value
+    // Note: the `std::sync::Mutex` is used intentionally, since contention never actually happens
+    // and we want to avoid cancelation because of it
+    val_rx: Arc<std::sync::Mutex<Option<tokio::sync::oneshot::Receiver<T>>>>,
+    val: sync::OnceCell<T>,
+}
+
+impl<T> Jit<T>
+where
+    T: MaybeSend,
+{
+    /// Create `Jit` value, and spawn a future `f` that computes its value
+    ///
+    /// Unlike normal Rust futures, the `f` executes eagerly (is spawned as a
+    /// tokio task).
+    pub fn new<Fut>(f: impl FnOnce() -> Fut + 'static + MaybeSend) -> Self
+    where
+        Fut: Future<Output = T> + 'static + MaybeSend,
+        T: 'static,
+    {
+        let (cancel_tx, mut cancel_rx) = tokio::sync::mpsc::channel(1);
+        let (val_tx, val_rx) = oneshot::channel();
+        super::imp::spawn(
+            &format!("Jit {} value", std::any::type_name::<T>()),
+            async move {
+                select! {
+                    _ = cancel_rx.recv() => {
+                        trace!(target: LOG_TASK, r#type = %std::any::type_name::<T>(), "Jit value future canceled");
+                    },
+                    val = f() => {
+                        match val_tx.send(val) {
+                            Ok(_) => {
+                                trace!(target: LOG_TASK, r#type = %std::any::type_name::<T>(), "Jit value ready");
+                            },
+                            Err(_) => {
+                                trace!(target: LOG_TASK,  r#type = %std::any::type_name::<T>(), "Jit value ready, but ignored");
+                            },
+                        };
+                    },
+                }
+            },
+        )
+        .expect("spawn not fail");
+
+        Self {
+            _cancel_tx: cancel_tx,
+            val_rx: Arc::new(Some(val_rx).into()),
+            val: sync::OnceCell::new(),
+        }
+    }
+
+    /// Get the reference to the value, potentially blocking for the
+    /// initialization future to complete
+    pub async fn get(&self) -> &T {
+        #[allow(clippy::await_holding_lock)]
+        self.val
+            .get_or_init(|| async {
+                self.val_rx
+                    // this lock gets locked only once so it's kind of useless other than making
+                    // Rust happy, but the overhead doesn't matter
+                    .lock()
+                    .expect("locking can't fail")
+                    .as_mut()
+                    .expect("value take only once")
+                    .await
+                    .unwrap_or_else(|_| panic!("Jit value {} panicked", std::any::type_name::<T>()))
+            })
+            .await
+    }
+}
+
+/// A value that initializes eagerly in parallel in a falliable way
+#[derive(Debug, Clone)]
+pub struct JitTry<T> {
```

Comment:
I agree, storing a result doesn't seem that much of an overhead.

Reply:
`&Result<_>` is a pain, as `Error` is generally not-cloneable. That's why `JitTry` does something slightly different.

Reply:
maybe `Jit` can be `JitTry<T, Infailable>` then

Reply:
Will try.

Reply:
But then `fn get()` is still going to be returning `Result` ...

Reply:
And in `fn get()` I'd like to keep returning some `Err`, even after moving out of the original one. Because of that I need to specialize over `anyhow::Result`, and can't do it generic over `std::result::Result`.

Hmmm...